### DISCUSSION TITLE
Basic example of how it works on dotnet maui

### DISCUSSION THREAD

**Discussion by mateusfrancino** [TolikPylypchuk/SharpHook#95] [Q&A] [answered]
I tried to use the package in a real case, but I couldn't understand how to use it, I would like to see a basic real case of how to use it, for example, if in the standard project that is created in dotnet where there is a counter, how do I do it? that when my user presses the enter key on the keyboard, I call the button click method?

```
namespace ExampleKeysIndiFarma
{
    public partial class MainPage : ContentPage
    {
        int count = 0;

        public MainPage()
        {
            InitializeComponent();
        }

        private void OnCounterClicked(object sender, EventArgs e)
        {
            count++;

            if (count == 1)
                CounterBtn.Text = $"Clicked {count} time";
            else
                CounterBtn.Text = $"Clicked {count} times";

            SemanticScreenReader.Announce(CounterBtn.Text);
        }
    }
}
```

**Comment by TolikPylypchuk** [comment, marked as answer]
Hello! Thanks for asking!

Please, bear in mind that SharpHook lets you create hooks that are **global to the entire OS**. The hook will work even if the application is not in the foreground, and it requires [special permissions on macOS](https://sharphook.tolik.io/articles/os-constraints.html#macos). Are you sure that this is what you really need? Perhaps you need to handle keyboard events only when your application is in the foreground, or you need to handle the <kbd>Enter</kbd> key only when a button is in focus - and for that you should use MAUI-specific APIs.

If you really do need a global hook, then here's a basic usage example.

To be honest, I don't have any experience with MAUI, but the basic idea is that you create an instance of `IReactiveGlobalHook` (e.g. `TaskPoolGlobalHook`), add event handlers to it, and run it. That's it!

The hook object should be global and should in general exist for the whole time your application runs (or not, depending on your use-case).

In your example this can be achieved like this:

```c#
namespace ExampleKeysIndiFarma
{
    public partial class MainPage : ContentPage
    {
        int count = 0;

        public MainPage()
        {
            InitializeComponent();

            var hook = new TaskPoolGlobalHook();
            hook.KeyPressed += (sender, e) =>
            {
                if (e.Data.KeyCode == KeyCode.VcEnter)
                {
                    this.Dispatcher.Dispatch(() => OnCounterClicked(sender, e));
                }
            };

            _ = hook.RunAsync();
        }

        private void OnCounterClicked(object sender, EventArgs e)
        {
            count++;

            if (count == 1)
                CounterBtn.Text = $"Clicked {count} time";
            else
                CounterBtn.Text = $"Clicked {count} times";

            SemanticScreenReader.Announce(CounterBtn.Text);
        }
    }
}
```

Don't forget to wrap stuff that modifies UI into `Dispatcher.Dispatch` since global hook event handlers don't run on the UI thread.

Remember that there must be **one global hook in the entire application**, so you should probably create it in the application class or something like that and then pass it down to individual views if needed. The example above is too simple to use in a real application.

Here's another example which uses dependency injection:

```c#
// MauiProgram.cs

public static class MauiProgram
{
    public static MauiApp CreateMauiApp()
    {
        var builder = MauiApp.CreateBuilder();
        builder.UseMauiApp<App>();

        builder.Services.AddSingleton<IGlobalHook>(sp => new TaskPoolGlobalHook(runAsyncOnBackgroundThread: true));

        var app = builder.Build();

        var hook = app.Services.GetRequiredService<IGlobalHook>();
        hook.RunAsync();

        return app;
    }
}

// MainPage.xaml.cs

public partial class MainPage : ContentPage
{
    int count = 0;

    public MainPage()
    {
        this.InitializeComponent();

        // There are most probably better ways to get the global hook from the service provider, this is just for simplicity
        var hook = Application.Current!.Handler.MauiContext!.Services.GetRequiredService<IGlobalHook>();

        this.Loaded += (sender, e) => hook.KeyPressed += OnKeyPressed;
        this.Unloaded += (sender, e) => hook.KeyPressed -= OnKeyPressed;
    }

    public void OnKeyPressed(object? sender, KeyboardHookEventArgs e)
    {
        if (e.Data.KeyCode == KeyCode.VcEnter)
        {
            this.Dispatcher.Dispatch(() => OnCounterClicked(sender, e));
        }
    }

    private void OnCounterClicked(object? sender, EventArgs e)
    {
        count++;

        if (count == 1)
            CounterBtn.Text = $"Clicked {count} time";
        else
            CounterBtn.Text = $"Clicked {count} times";

        SemanticScreenReader.Announce(CounterBtn.Text);
    }
}
```

**Reply by mateusfrancino** [reply]
Well, first of all, thank you. Because your code clarified a lot of things for me. Regarding the question of whether I need to activate the keyboard shortcut globally, to be honest, I am creating a POS system, so for example on the main screen when I press the F2 key, the software should redirect me to another screen, but this is the same shortcut can be used on another screen for another purpose, in the same way as enter, on one screen it can be used to perform that task for that screen, but when I call it on another screen, it must perform another situation according to the which I configure, so that he is always looking at the screen that is in the foreground. And apply the rules defined for that screen. From the way you explained it, I understood that when I define a shortcut it will always call the same thing, even if the screen is in 2nd plane. **THIS IS NOT WHAT I NEED**, in my case the shortcuts must be dynamic.

**Comment by mateusfrancino** [comment]
How could I create key combination, for example ctrl + enter performs such action

**Reply by TolikPylypchuk** [reply]
SharpHook doesn't provide such functionality, you will have to implement it yourself. It's simple to do with SharpHook.Reactive, but only if you are already familiar with Rx.NET.

You can take a look at [this code in my other project](https://github.com/TolikPylypchuk/KeyboardSwitch/blob/bcbf1247c54240f2988146c4c3989d8db9cc7a1e/src/KeyboardSwitch.Core/Services/Hook/SharpHookService.cs) which uses SharpHook to handle key combinations. It works with `ModifierMask` since it only needs to handle modifier key combinations, but it could be changed to work with `KeyCode` instead.
